src/main/java/com/redhat/devtools/intellij/lsp4mp4ij/psi/internal/graphql/java/MicroProfileGraphQLASTValidator.java

```diff
@@ -14,65 +14,212 @@
+ *
  * @see https://download.eclipse.org/microprofile/microprofile-graphql-1.0/microprofile-graphql.html
  */
 public class MicroProfileGraphQLASTValidator extends JavaASTValidator {
 
-	private static final Logger LOGGER = Logger.getLogger(MicroProfileGraphQLASTValidator.class.getName());
-
-	private static final String NO_VOID_MESSAGE = "Methods annotated with microprofile-graphql's `@Query` cannot have 'void' as a return type.";
-	private static final String NO_VOID_MUTATION_MESSAGE = "Methods annotated with microprofile-graphql's `@Mutation` cannot have 'void' as a return type.";
-	@Override
-	public boolean isAdaptedForDiagnostics(JavaDiagnosticsContext context) {
-		Module javaProject = context.getJavaProject();
-		// Check if microprofile-graphql is on the path
-		return PsiTypeUtils.findType(javaProject, MicroProfileGraphQLConstants.QUERY_ANNOTATION) != null;
-	}
-
-	@Override
-	public void visitMethod(PsiMethod node) {
-		validateMethod(node);
-		super.visitMethod(node);
-	}
-
-	private void validateMethod(PsiMethod node) {
-		// ignore constructors, and non-void methods for now, it's faster than iterating through all annotations
-		if (node.getReturnTypeElement() == null ||
-			!PsiType.VOID.equals(node.getReturnType())) {
-			return;
-		}
-		for (PsiAnnotation annotation : node.getAnnotations()) {
-			if (isMatchAnnotation(annotation, MicroProfileGraphQLConstants.QUERY_ANNOTATION) ) {
-				super.addDiagnostic(NO_VOID_MESSAGE, //
-						MicroProfileGraphQLConstants.DIAGNOSTIC_SOURCE, //
-						node.getReturnTypeElement(), //
-						MicroProfileGraphQLErrorCode.NO_VOID_QUERIES, //
-						DiagnosticSeverity.Error);
-			} else if (isMatchAnnotation(annotation, MicroProfileGraphQLConstants.MUTATION_ANNOTATION)) {
-				super.addDiagnostic(NO_VOID_MUTATION_MESSAGE, //
-						MicroProfileGraphQLConstants.DIAGNOSTIC_SOURCE, //
-						node.getReturnTypeElement(), //
-						MicroProfileGraphQLErrorCode.NO_VOID_MUTATIONS, //
-						DiagnosticSeverity.Error);
-			}
-		}
-	}
+    private static final Logger LOGGER = Logger.getLogger(MicroProfileGraphQLASTValidator.class.getName());
+
+    private static final String NO_VOID_MESSAGE = "Methods annotated with microprofile-graphql's `@Query` cannot have 'void' as a return type.";
+    private static final String NO_VOID_MUTATION_MESSAGE = "Methods annotated with microprofile-graphql's `@Mutation` cannot have 'void' as a return type.";
+    private static final String WRONG_DIRECTIVE_PLACEMENT = "Directive ''{0}'' is not allowed on element type ''{1}''";
+
+    boolean allowsVoidReturnFromOperations = true;
+
+
+    @Override
+    public boolean isAdaptedForDiagnostics(JavaDiagnosticsContext context) {
+        Module javaProject = context.getJavaProject();
+        if(PsiTypeUtils.findType(javaProject, MicroProfileGraphQLConstants.QUERY_ANNOTATION) == null) {
+            return false;
+        }
+        String quarkusVersion = QuarkusModuleUtil.getQuarkusVersion(context.getJavaProject());
```

Comment:
For the moment we don't need the getQuarkusVersion, we need the method `isQuarkusVersion(Module javaProject, String version)`

So I suggest:

 * set getQuarkusVersion as private
 * introduce a new new method `isQuarkusVersion(Module javaProject, String version)`

So you could just write after that:

`boolean allowsVoidReturnFromOperations = QuarkusModuleUtil.isQuarkusVersion(context.getJavaProject(), "3.1"));`

Reply:
I'm not following here, we need to check a range (`<3.1`), not one particular version number, so with `isQuarkusVersion` we would have to list all versions after 3.1, many of which aren't known yet. 
Even if we inverted the condition and then enumerated all versions BEFORE 3.1, we can't be 100 % sure that there won't ever be a 2.17 release, for example.

Reply:
> I'm not following here, we need to check a range (<3.1)

You are right, sorry I read too quickly your code. So My suggestion was to have an utility method like 

`QuarkusModuleUtil.isQuarkusVersionGreaterThan(String version)`  I'm pretty sure that we will need this utility method for another features.

Reply:
And what should that return if we can't detect the version at all (there might not be a Quarkus dependency at all) or it's a custom version, like 999-SNAPSHOT, where we can't reason about anything? I assume both of these cases should return false, right?

Reply:
One more issue is how do you define "greater than"?  It might not be a sufficient determinator in all cases
For example:
- 4.0 does not have a feature
- 4.1 has it
- 3.0 doesn't have it
- 3.1 has it

Then, how would you check whether a particular version has this feature, using `isQuarkusVersionGreaterThan`? You can't use `isQuarkusVersionGreaterThan(3.1)` because then it would return `true` for 4.0, which is wrong. 
This approach, while it is convenient, has a lot of limitations IMO...

Reply:
I understand what you mean, but it would be very nice to have an utility method which check your usecases with version. I would like to avoid copy/paste code 

```
if(quarkusVersion != null) {
            Matcher matcher = QuarkusModuleUtil.QUARKUS_STANDARD_VERSIONING.matcher(quarkusVersion);
            if(matcher.matches()) { // ignore if this is not a standard version (for example, a custom snapshot)
```

for an another check version for another feature.

So I sggest that you define a method like

`boolean checkQuarkusVersion(Project project, Predicate<Matcher> versionMatcher)`

After that you could use it like this:

```
checkQuarkusVersion(context.getJavaProject(), matcher -> Integer.valueOf(matcher.group(1)) < 3 ||
                        (Integer.valueOf(matcher.group(1)) == 3) && Integer.valueOf(matcher.group(2)) == 0));
```

Reply:
Ok that makes sense, I'll do that in a bit

Reply:
Done, please check

Reply:
It is perfect, Thanks!